DSI updates to use the DSI0 and JDI LT070ME05000 panel

6by9 opened the pull request:
Still not 100% reliable :-(

Fixes up DSI0 with the mainline patches rather than earlier versions.

mripard committed:
drm/vc4: drv: Remove the DSI pointer in vc4_drv

Commit 51f4fcd9c4ea867c3b4fe58111f342ad0e80642a upstream.

That pointer isn't used anywhere, so there's no point in keeping it.

Reviewed-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Link: https://patchwork.freedesktop.org/patch/msgid/20201203132543.861591-2-maxime@cerno.tech

mripard committed:
drm/vc4: dsi: Use snprintf for the PHY clocks instead of an array

Commit dc0bf36401e891c853e0a25baeb4e0b4e6f3626d upstream.

The DSI clocks setup function has been using an array to store the clock
name of either the DSI0 or DSI1 blocks, using the port ID to choose the
proper one.

Let's switch to an snprintf call to do the same thing and simplify the
array a bit.

Reviewed-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Link: https://patchwork.freedesktop.org/patch/msgid/20201203132543.861591-4-maxime@cerno.tech

mripard committed:
drm/vc4: dsi: Introduce a variant structure

Commit d1d195ce26a14ec0a87816c09ae514e1c40e97f7 upstream.

Most of the differences between DSI0 and DSI1 are handled through the
ID. However, the BCM2711 DSI is going to introduce one more variable to
the mix and will break some expectations of the earlier, simpler, test.

Let's add a variant structure that will address most of the differences
between those three controllers.

Reviewed-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Link: https://patchwork.freedesktop.org/patch/msgid/20201203132543.861591-5-maxime@cerno.tech

6by9 committed:
drm/vc4: dsi: Add support for DSI0

Commit 4b265fe11fad4234b12d92dd8091f9aa0c878eea upstream.

DSI0 was partially supported, but didn't register with the main
driver, and the code was inconsistent as to whether it checked
port == 0 or port == 1.

Add compatible string and other support to make it consistent.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Link: https://patchwork.freedesktop.org/patch/msgid/20201203132543.861591-6-maxime@cerno.tech

6by9 committed:
drm/vc4: dsi: Add configuration for BCM2711 DSI1

Commit d0666be8ef9e8e65d4b7fabc1606ec51f61384c0 upstream.

BCM2711 DSI1 doesn't have the issue with the ARM not being
able to write to the registers, therefore remove the DMA
workaround for that compatible string.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
Signed-off-by: Maxime Ripard <maxime@cerno.tech>
Link: https://patchwork.freedesktop.org/patch/msgid/20201203132543.861591-8-maxime@cerno.tech

6by9 committed:
drm/atomic: Don't fixup modes that haven't been reset

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
drm/vc4: Register dsi0 as the correct vc4 encoder type

vc4_dsi was registering both dsi0 and dsi1 as VC4_ENCODER_TYPE_DSI1
which seemed to work OK for a single DSI display, but fails
if there are two DSI displays connected.

Update to register the correct type.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
drm/vc4: Fix dsi0 interrupt support.

DSI0 seemingly had very little or no testing as a load of
the register mappings were incorrect/missing, so host
transfers always timed out due to enabling/checking incorrect
bits in the interrupt enable and status registers.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
drm/vc4: Correct pixel order for DSI0

For slightly unknown reasons, dsi0 takes a different pixel format
to dsi1, and that has to be set in the pixel valve.

Amend the setup accordingly.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
Revert "drm/vc4: Add configuration for BCM2711 DSI1."

This reverts commit 10faa28e1474e6cf33cb0809d2a6d5b7a8351a28.

Revert in order to take the cleaner upstream version.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
Revert "drm/vc4: Add support for DSI0"

This reverts commit d704a3c965783ed51476031206b2120676418672.

Revert in order to apply upstream version

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

pelwell commented on drivers/gpu/drm/vc4/vc4_dsi.c:
```diff
@@ -1060,12 +1105,9 @@ static void vc4_dsi_encoder_enable(struct drm_encoder *encoder)
-	}
+	DSI_PORT_WRITE(PHY_AFEC0,
+		       DSI_PORT_READ(PHY_AFEC0) &
+		       ~DSI1_PHY_AFEC0_RESET);
```
DSI0_PHY_AFEC0_RESET and DSI1_PHY_AFEC0_RESET are different bit positions (in different registers, although the interleaved definitions don't make that obvious). The reset code above uses the correct value for the port in question, but here the DSI1 value is used in both cases. Use `DSI_PORT_BIT(PHY_AFEC0_RESET)` instead?

pelwell commented on drivers/gpu/drm/vc4/vc4_dsi.c:
```diff
@@ -1184,13 +1226,28 @@ static ssize_t vc4_dsi_host_transfer(struct mipi_dsi_host *host,
-	if (msg->rx_len) {
-		DSI_PORT_WRITE(INT_EN, (DSI1_INTERRUPTS_ALWAYS_ENABLED |
-					DSI1_INT_PHY_DIR_RTF));
+	if (!dsi->port) {
```
`dsi->port == 0` is used above (but not added in this commit).

pelwell commented on arch/arm/boot/dts/overlays/vc4-kms-dsi-7inch-overlay.dts:
```diff
@@ -6,51 +6,97 @@
 		__overlay__ {
-			#address-cells = <1>; size-cells = <0>;
+			#address-cells = <1>;
+			size-cells = <0>;
```
`#size-cells`

pelwell commented on arch/arm/boot/dts/overlays/edt-ft5406.dtsi:
```diff
@@ -0,0 +1,42 @@
+ */
+
+/dts-v1/;
+/plugin/;
```
I'm not keen on having these tags in a dtsi file - they should only appear once in a DTS/overlay file, and I feel they're better of in the parent .dts (although there's at least one counter-example).

pelwell replied:
Alternatively, given that the .dtsi here is a complete overlay, I'd be happy with it remaining as a .dts, even though it is included by another overlay (although the including file should then lose it's own header tags).

pelwell commented on arch/arm/boot/dts/overlays/README:
```diff
@@ -3141,10 +3141,14 @@ Params: cma-512                 CMA is 512MB (needs 1GB)
-        Use edt-ft5406 for the touchscreen element.
+        Includes the edt-ft5406 for the touchscreen element.
         Requires vc4-kms-v3d to be loaded.
 Load:   dtoverlay=vc4-kms-dsi-7inch
```
Sorry, but this now needs `,<param>=<val>`.

pelwell commented on arch/arm64/configs/bcmrpi3_defconfig:
```diff
@@ -708,6 +708,7 @@ CONFIG_STMPE_SPI=y
 CONFIG_MFD_ARIZONA_I2C=m
 CONFIG_MFD_ARIZONA_SPI=m
 CONFIG_MFD_WM5102=y
+CONFIG_REGULATOR_RASPBERRYPI_TOUCHSCREEN_ATTINY=m
```
This setting isn't working - it needs at least CONFIG_REGULATOR=y, but that doesn't seem to be sufficient.

6by9 committed:
drm/vc4: Add correct stop condition to vc4_dsi_encoder_disable iteration

vc4_dsi_encoder_disable is partially an open coded version of
drm_bridge_chain_disable, but it missed a termination condition
in the loop for ->disable which meant that no post_disable
calls were made.

Add in the termination clause.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
dtoverlays: Add overlays for JDI LT070ME05000 1200x1920 DSI panel

Credit to forum member gizmomouse on
https://www.raspberrypi.org/forums/viewtopic.php?f=98&t=253912 and
Andrey Vostrukhin of Harlab for these overlays.

See https://github.com/harlab/CM4_LCD_LT070ME05000 for
schematics and docs for the adapter board to connect this panel which
is found in the Asus/Google 2013 Nexus 7" tablet and therefore
relatively easily available.

Note that this uses 4 DSI data lanes, and therefore MUST be used
with DISP1 on a Compute Module. It can not be used on a standard
Pi.

There are two versions of the adapter board. V1 connects the
display controls to Pi GPIOs, whilst v2 uses an I2C GPIO expander
so needs no additional connections beyond the FFC and power.

The touchscreen overlay for these panels varies, so that part
is not configured.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
drm/panel: jdi-lt070me05000: Use gpiod_set_value_cansleep

There is no reason why the control GPIOs for the panel can not
be connected to I2C or similar GPIO interfaces that may need to
sleep, therefore switch from gpiod_set_value to
gpiod_set_value_cansleep calls to configure them.
Without that you get complaints from gpiolib every time the state
is changed.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
defconfigs: Add PANEL_JDI_LT070ME05000 DSI panel

Used by the 2013 Nexus 7" tablet, and working via an adapter
board on Compute Modules.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
Rebased, updated based on your comments.

I've dropped the 7inch panel updates as I haven't had a chance to resolve the issues, but added in the overlay for the JDI LT070ME05000 panel that is now working using the hardware documented at https://github.com/harlab/CM4_LCD_LT070ME05000

6by9 commented:
@harlab Are you happy with the text I've added for your contributions?

harlab commented:
> @harlab Are you happy with the text I've added for your contributions?

@6by9 overlays were based on work of Pi forum user, so I can’t take full credit. Leaving only repo link should be fine. 
Thanks for your work.

6by9 commented:
OK, will drop your email address (avoids some spam!)

Just so that I can include correct credits, was it gizmomouse at https://www.raspberrypi.org/forums/viewtopic.php?f=98&t=253912 that you're referencing? I'm happy to include a credit to them too.

harlab commented:
> OK, will drop your email address (avoids some spam!)

Thanks

> Just so that I can include correct credits, was it gizmomouse at https://www.raspberrypi.org/forums/viewtopic.php?f=98&t=253912 that you're referencing? I'm happy to include a credit to them too.

@6by9 yes, that’s what overlays are based on.

6by9 commented:
Updated commit text based on harlab's comments, and I believe addressed all of pelwell's comments on the earlier version.

Hopefully all good now with that display, so now I can go back to the Pi 7" one.

pelwell commented on arch/arm/boot/dts/overlays/README:
```diff
@@ -3158,6 +3158,27 @@ Load:   dtoverlay=vc4-kms-dsi-7inch
+Load:   dtoverlay=vc4-kms-dsi-lt070me05000,<param>
+Params: reset                   GPIO for the reset signal (default "17")
+        enable                  GPIO for the enable signal (default "4")
+        dcdc-en                 GPIO for the DC-DC converter enable (default "5")
```
This line is one character too long. I would drop the quotes from the defaults.

6by9 replied:
Done.
We obviously need to get you a bigger monitor as Linus has relented on the hard limit of 80 columns - https://lkml.org/lkml/2020/5/29/1038

pelwell replied:
I became aware of the increased line length limit last week. I still prefer the non-string defaults without quotes, but I could relax the 80-char limit in the script a bit (after checking that the dtoverlay/dtparam utils won't die horribly).

pelwell replied:
```
    char line_buf[82];
```
Some work required there.

pelwell reviewed (commented):
One last niggle.
